레스토랑 상세 페이지 구현하기(라우팅 추가하기)

yoonhe opened the pull request:
코드리뷰 요청드립니다 😆

## 한 일
- `about`, `restaraunts`, `restaraunts/:id` 라우팅 작업

## 아쉬운점
- `RestaurantsDetailPage`에 너무 많은 것을 담아버렸습니다.. 내일 좀더 분리하도록 할게요!

## 디엠으로 질문한 내용 정리
### 질문 1
restaurants/1 로 페이지 이동되었을때 `RestaurantsDetailContainer`에서 `getParams`를 이용해 `param`을 얻어와 레스토랑 상세 데이터를 가져오는 `api 요청`에 사용해주기위해 아래와같이 코드를 구현해주었는데요

```js
const { id } = useParams();
useEffect(() => {
    dispatch(loadRestaurantDetail(id));
}, []);
```

`RestaurantsDetailContainer.test` 에서 `dispatch` 에 대한 테스트만 하면 될지.. 아니면  `useParams`에 대한 테스트도 추가해야하나 고민중인데요.. `react-router-dom` 에서 제공하는 함수라 `useSelector`를 테스트할때 `mocking` 하는 것처럼 해당 함수에서 실제로 `restaurants/1` 해당 경로로 진입했을때 `param`으로 `1`이라는 값을 가져오는지에 대한 테스트가 필요없다고 생각하는데...
이미 누군가가 잘 만들어서 배포한 라이브러리에서 제공하는 함수를 믿고 사용하는게 맞나요? 아니면 테스트를 해주는게 맞는건가요?

#### 질문 1 답변

라이브러리에 대한 테스트까지는 할 필요가 없죠 :)
따로 해주고 싶다면 `useParam` 말고 `match`를 써주는 방법이 있긴한데 어차피 메모리 라우터 써주는거면 안해도 될것 같아요

---

Commit:
[red] RestaurantsPage에 대한 테스트를 추가합니다

Commit:
[green] RestaurantsPage를 구현합니다

Commit:
react-router-dom을 설치합니다

Commit:
[red] App에 대한 테스트를 추가합니다.

path가 '/'일 경우에 대한 라우팅 테스트를 추가합니다

Commit:
[green] App을 구현합니다

react-router-dom을 사용하여 path가 '/'인 라우팅 구현

Commit:
[refactoring] App을 리팩토링한다

Commit:
[red] App에 대한 테스트코드를 추가합니다

'/about'을 위한 라우팅 테스트입니다

Commit:
[green] App을 구현합니다

Commit:
[red] App에 대한 테스트를 추가합니다

header에 대한 테스트를 추가합니다

Commit:
[green] App을 구현합니다

Commit:
[green] App을 리팩토링합니다

Commit:
[red] App에 대한 테스트코드를 추가합니다

'/restaurants' 라우팅 테스트를 추가합니다

Commit:
[green] App을 구현합니다
'/restaurants' 라우팅을 구현합니다

Commit:
[red] App에 대한 테스트를 추가합니다

존재하지않는 path에 대한 라우팅 테스트를 추가합니다

Commit:
[green] App을 구현합니다

존재하지않는 path에 대한 라우팅을 구현합니다

Commit:
[red] AboutPage에 대한 테스트를 추가합니다

Commit:
[green] AboutPage를 구현합니다

Commit:
[red] NotFoundPage에 대한 테스트를 추가합니다

Commit:
[green]] NotFoundPage를 구현합니다

Commit:
[red] HomePage에 대한 테스트를 추가합니다

Commit:
[green] HomePage를 구현합니다

Commit:
[refactoring] App을 리팩토링합니다

Commit:
[red] App에 대한 테스트를 추가합니다

'/restaurants/:id'에 대한 라우팅을 테스트합니다

Commit:
[green] App을 구현합니다.

Commit:
[red] RestaurantsDetailePage에 대한 테스트를 추가한다

Commit:
[green] RestaurantsDetailePage를 구현한다

Commit:
파일 이름 변경

Commit:
[red] RestaurantsDetailPage에 대한 테스트를 추가합니다

Commit:
[green] RestaurantsDetailPage를 구현합니다

Commit:
[red] actions에 대한 테스트를 추가합니다.

loadRestaurantDetail action

Commit:
[green] action을 구현합니다
loadRestaurantDetail action

Commit:
[refactoring] action을 리팩토링합니다

loadRestaurantDetail action

Commit:
[red] api에 대한 테스트코드를 추가합니다
fetchRestaurantDetail api

Commit:
[green] api를 구현합니다
fetchRestaurantDetail api

Commit:
[red] reducer에 대한 테스트를 추가합니다

setRestaurantDetail reducer

Commit:
[green] reducer를 구현합니다

Commit:
테스트 데이터 추가(restaurantDetail)

Commit:
[refactoring] RestaurantsPage를 리팩토링합니다

Commit:
[refactoring] App을 리팩토링합니다

Commit:
[refactoring] RestaurantsContainer를 리팩토링합니다

버튼 클릭시 라우팅될 수 있도록 `Link`로 감싸주었습니다

Commit:
[refactoring] RestaurantsDetailPage를 리팩토링합니다

yoonhe commented:
질문이 있습니다!

`restaurants/:id`를 제외한 나머지 페이지들(about, restaurents)은 새로고침을 해도 페이지가 잘 나오는데 ``restaurants/:id` 페이지만 새로고침시에 아래와 같은 에러가납니다.
![image](https://user-images.githubusercontent.com/53102889/109645482-3f2c9b00-7b9a-11eb-8d9a-1022bc619d20.png)

다른페이지들과 다른부분을 찾아보니
`http://localhost:8080/restaurants/main.js` 이렇게 `main.js`를 다른 경로에서 가져와 에러가 나는것 같습니다.. 원인은 파악했는데 해결방법을 못찾고있습니다..

moonkii commented on src/AboutPage.jsx:
```diff
@@ -0,0 +1,12 @@
+  );
+}
+
+export default AboutPage;
```
따로 분리하지말고 `export default function AboutPage()...`  로 합쳐주면 좋을 것 같아요.

moonkii commented on src/AboutPage.test.jsx:
```diff
@@ -0,0 +1,11 @@
+
+import AboutPage from './AboutPage';
+
+test('AboutPage', () => {
+  render((
+    <AboutPage />
+  ));
+});
```
간단하지만 제목이라도 나오도록 테스트를 넣어주면 어떨까요? 🤔

moonkii commented on src/RestaurantsContainer.test.jsx:
```diff
@@ -4,18 +4,25 @@ import { render } from '@testing-library/react';
 
 import { useSelector } from 'react-redux';
 
+import { MemoryRouter } from 'react-router-dom';
 import RestaurantsContainer from './RestaurantsContainer';
```
`import` 도 공백으로 구분지어주면 좋을 것 같아요.

moonkii commented on src/RestaurantsDetailPage.jsx:
```diff
@@ -0,0 +1,37 @@
+
+  useEffect(() => {
+    dispatch(loadRestaurantDetail(id));
+  }, [id]);
```
load 하는 시점은 id 값이 변할 때가 좋을까요? 🤔

yoonhe replied:
깃헙에 슬랙 내용을 기록해두려고 가져왔습니다 ㅎㅎㅎㅎ

상세 페이지는 레스토랑 메뉴(마녀주방 등등..)을 클릭했을때 라우팅되거나 주소창에 직접 입력했을때 보여지므로 `id`가 바뀌어 리렌더링될 상황이 생기지않으므로 두번쨰 인자에 `id` 를 넣어줄 필요가 없는거군요?? 두번쨰 인자에 빈배열을 넣어 최초 렌더링시에만 `useEffect` 가 실행되게 하면 된다고 이해했는데, 틀린 부분이 있으면 말씀해주세요!

#### @moonkii 님 답변
네 맞아요!

moonkii commented on src/RestaurantsDetailPage.test.jsx:
```diff
@@ -0,0 +1,20 @@
+import { useDispatch } from 'react-redux';
+import RestaurantsDetailPage from './RestaurantsDetailPage';
+
+test('RestaurantsDetailPage', () => {
+  const dispatch = jest.fn();
+
+  useDispatch.mockImplementation(() => dispatch);
+
+  render((
+    <MemoryRouter>
+      <RestaurantsDetailPage />
+    </MemoryRouter>
+  ));
+});
```
왜 테스트 코드는 없나요??

yoonhe replied:
`RestaurantsDetailPage`를 마크업부터 다시 `TDD` 해보았습니다. 관련내용은 푸쉬해두었습니다 !

moonkii reviewed (changes requested):
![image](https://user-images.githubusercontent.com/46776329/109673401-575fe280-7bb9-11eb-9a23-58396da1e758.png)

TDD 를 하실 때 `RED` 에서 확인하고 싶은 내용이 테스트로 들어가야해요.
테스트 없이 먼저 작성되는 코드들이 많은 것 같아요!

> ... main.js를 다른 경로에서 가져와 에러가 나는것 같습니다..

네 원인은 잘 파악하신 것 같아요.
base 가 되는 path 가 다르기 때문에 발생하는 문제에요.
해결 방법으로는 `main.js` 의 경로를 잡아주는 방법도 있고, `webpack.config.js` 에서 옵션을 주는 방법이 있을거에요.

Commit:
main.js 경로 변경

상세페이지 라우팅후 새로고침시 main.js가 경로를 제대로 잡지 못해 404가 발생하여
`main.js`에서 `/main.js`로 변경해주었습니다

Commit:
[red] RestaurantsDetailPage에 대한 테스트코드를 추가합니다.

테스트코드 없이 작성된 코드들이 많은것 같아 다시 마크업부터 테스트코드를
추가해보았습니다

Commit:
[green] RestaurantsDetailPage에 대한 테스트코드를 추가합니다.

Commit:
[red] RestaurantsDetailContainer에 대한 테스트를 추가합니다

Commit:
[green] RestaurantsDetailContainer를 구현합니다

Commit:
[red] RestaurantsDetailContainer에 대한 테스트를 추가합니다

dispatch 테스트

Commit:
[green] RestaurantsDetailContainer를 구현합니다

Commit:
[refactoring] RestaurantsDetailPage를 리팩토링합니다

Commit:
given2 라이브러리 설치

yoonhe commented:
> 네 원인은 잘 파악하신 것 같아요.
base 가 되는 path 가 다르기 때문에 발생하는 문제에요.
해결 방법으로는 main.js 의 경로를 잡아주는 방법도 있고, webpack.config.js 에서 옵션을 주는 방법이 있을거에요.

슬랙에 효근님께서 올려주신 아래의 질문글에 대한 답변에서 답을 찾을 수 있었는데요
https://codesoom-group.slack.com/archives/C01JEHR319T/p1614692515012800

아래처럼 `/main.js`로 경로가 잡히면 절대경로가 되기때문에 어떤 페이지에서 `main.js `를 가져오던 `루트경로/main.js`에서 가져오게되서 `main.js`가 올바르게 가져와진다고 이해했는데, 

잘못 이해한 부분이 있으면 말씀해주세요!
```js
<script src="/main.js"></script>
```

Commit:
[red] RestaurantsDetail에 대한 테스트를 추가합니다

Commit:
[green]] RestaurantsDetail을 구현합니다

Commit:
[refactoring] RestaurantsDetail을 리팩토링합니다

Commit:
[refactoring] AboutPage를 리팩토링합니다

Commit:
[refactoring] HomePage를 리팩토링합니다

Commit:
[refactoring] RestaurantsContainer를 리팩토링합니다

Commit:
[refactoring] RestaurantsPage를 리팩토링합니다

Commit:
RestaurantsDetail이 포함된 파일명 일부를 RestaurantDetail으로 변경

하나의 레스토랑에 대한 상세페이지이기 떄문에 `Restaurants`에서 `Restaurant`으로
변경해주었습니다

Commit:
[refactoring] App을 리팩토링합니다
App 테스트코드를 리팩토링합니다

Commit:
[refactoring] NotFoundPage를 리팩토링합니다
NotFoundPage 테스트코드를 리팩토링합니다

Commit:
[refactoring] RestaurantDetailPage를 리팩토링합니다

moonkii commented on src/RestaurantDetail.jsx:
```diff
@@ -0,0 +1,24 @@
+            }
+          </dl>
+        </div>
+      ) : <div>로딩중..</div>}
```
컴포넌트를 `( )` 로 감싸주면 좋을 것 같아요.

린트에서는 안잡히나요?

yoonhe replied:
흠.. 넵..

```js
: (
        <div>로딩중..</div>
      )
```

이렇게 변경해주었습니다!

moonkii commented on src/RestaurantDetailContainer.test.jsx:
```diff
@@ -0,0 +1,55 @@
+    ));
+  }
+
+  beforeEach(() => {
```
mock clear 도 추가해주면 좋겠죠?

yoonhe replied:
넵! `jest.clearAllMocks();` 추가해주었습니다!

moonkii commented on src/RestaurantDetailContainer.test.jsx:
```diff
@@ -0,0 +1,55 @@
+  beforeEach(() => {
+    useDispatch.mockImplementation(() => dispatch);
+    useSelector.mockImplementation((selector) => selector({
+      restaurantDetail: given.restaurantDetail,
```
잘 사용하셨어요! 👍

moonkii commented on src/RestaurantDetailPage.jsx:
```diff
@@ -0,0 +1,12 @@
+
+function RestaurantDetailePage() {
+  return (
+    <>
+      <RestaurantDetailContainer />
+    </>
```
컴포넌트가 하나면 Fragment 로 감싸주지 않아도 괜찮아요!

yoonhe replied:
앗.. 제가 왜 그랬을까요 ㅎㅎㅎ 수정했습니다!

Commit:
[refactoring] RestaurantDetailPage를 리팩토링합니다

Commit:
[refactoring] RestaurantDetail를 리팩토링합니다

Commit:
[refactoring] RestaurantsDetailContainer를 리팩토링합니다

Commit:
[refactoring] RestaurantsDetailPage를 리팩토링합니다

moonkii commented on src/RestaurantDetailPage.jsx:
```diff
@@ -2,11 +2,7 @@ import React from 'react';
-      <RestaurantDetailContainer />
-    </>
-  );
+  return <RestaurantDetailContainer />;
```
`return ( );` 이 기본형태에요. 괄호를 넣어주면 좋을 것 같아요.

moonkii commented on src/RestaurantDetailContainer.jsx:
```diff
@@ -0,0 +1,26 @@
+    dispatch(loadRestaurantDetail(id));
+  }, []);
+
+  return (
+    <>
+      <RestaurantDetail restaurant={restaurantDetail} />
+    </>
+  );
```
https://github.com/CodeSoom/react-week6-assignment-1/pull/42#discussion_r587548537

코드 리뷰를 할 때 동일한 부분에 모두 리뷰를 달지는 않아요.

리뷰를 반영하실 때 똑같은 부분도 찾아서 전체적으로 반영해주면 좋을 것 같아요.

yoonhe replied:
넵... 좀더 꼼꼼히 살펴보로독 할게요!

moonkii commented on src/HomePage.test.jsx:
```diff
@@ -0,0 +1,15 @@
+
+import HomePage from './HomePage';
+
+test('HomePage', () => {
+  render((
+    <MemoryRouter>
+      <HomePage />
+    </MemoryRouter>
+  ));
+});
```
`HomePage` 테스트도 추가해주세요 ;)

moonkii commented:
PR을 `CodeSoom:main` 으로 보내셨어요!
올바른 branch 로  변경해서 보내주세요 ;)
